Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,161 @@
+//
+// `AsyncQueue` methods have particular expectations about whether they must be
+// invoked on the queue or not; check "preconditions" section in comments on
+// each method.
+//
+// A significant portion of `AsyncQueue` interface only exists for test purposes
+// and must *not* be used in regular code.
+class AsyncQueue {
+ public:
+  using Operation = internal::Executor::Operation;
+  // A more-or-less arbitrary unit of time for scheduling operations in the
+  // future.
+  using Milliseconds = internal::Executor::Milliseconds;
+
+  explicit AsyncQueue(std::unique_ptr<internal::Executor> executor);
+
+  // Asserts for the caller that it is being invoked asynchronously on the
+  // `AsyncQueue.`
+  void VerifyIsAsyncCall() const;
+  // Asserts for the caller that it is being invoked as part of an operation on
+  // the `AsyncQueue`.
+  void VerifyCalledFromOperation() const;
+
+  // Puts the `operation` on the queue to be executed as soon as possible, while
+  // maintaining FIFO order.
+  //
+  // Precondition: `Enqueue` calls cannot be nested; that is, `Enqueue` may not
+  // be called by a previously enqueued operation when it is run (as a special
+  // case, destructors invoked when an enqueued operation has run and is being
+  // destroyed may invoke `Enqueue`).
+  void Enqueue(const Operation& operation);
+
+  // Like `Enqueue`, but allowing nesting.
+  void EnqueueAllowingNesting(const Operation& operation);
+
+  // Puts the `operation` on the queue to be executed `delay` milliseconds from
+  // now, and returns a handle that allows to cancel the operation (provided it
+  // hasn't run already).
+  //
+  // `operation` is tagged by a `timer_id` which allows to identify the caller.
+  // Only one operation tagged with any given `timer_id` may be on the queue at
+  // any time; an attempt to put another such operation will result in an
+  // assertion failure. In tests, these tags also allow to check for presence of
+  // certain operations and to run certain operations in advance.
+  //
+  // Precondition: `EnqueueAfterDelay` is being invoked asynchronously on the
+  // queue.
+  DelayedOperation EnqueueAfterDelay(Milliseconds delay,
+                                     TimerId timer_id,
+                                     const Operation& operation);
+
+  // Immediately executes the `operation` on the queue.
+  //
+  // Precondition: the queue is idle at the moment of the call (no other
+  // operation is currently being executed).
+  //
+  // Precondition: `StartExecution` is being invoked asynchronously on the
+  // queue.
+  void StartExecution(const Operation& operation);
+
+  // Test-only interface follows
+
+  // Like `Enqueue`, but blocks until the `operation` is complete.
+  void EnqueueBlocking(const Operation& operation);
+
+  // Checks whether an operation tagged with `timer_id` is currently scheduled
+  // for execution in the future.
+  //
+  // Precondition: `StartExecution` is being invoked asynchronously on the
```

Comment:
I suspect you mean s/StartExecution/IsScheduled/

Reply:
Thanks for spotting this!